### PR TITLE
Fix Android shutdown error

### PR DESCRIPTION
**Pull request by Zeno-**
E.g.
Prior to patch:
```
V/minetest(16024): 2015-05-06 14:45:12: ERROR[main]: Shutting down minetest.
E/minetest(16024): Shutting down minetest.
I/ActivityManager(  783): Process net.minetest.minetest (pid 16024) (adj 0) has died.
I/WindowState(  783): WIN DEATH: Window{426a8c88 u0 net.minetest.minetest/net.minetest.minetest.MtNativeActivity}
W/ActivityManager(  783): Force removing ActivityRecord{4285eb28 u0 net.minetest.minetest/.MtNativeActivity t350}: app died, no saved state
```
After patch:
```
V/minetest( 2372): httpfetch_cleanup: cleaning up
V/minetest( 2372): 2015-05-06 17:32:59: ERROR[main]: Shutting down minetest.
E/minetest( 2372): Shutting down minetest
```

### CODE REVIEW

**Comment by Zeno-** [comment]
Issue: the application now **appears** to shutdown properly, but when trying to launch it a second time it will not launch (unless you close it from task manager first)

**Comment by Zeno-** [comment]
@sfan5, @sapier If either of you have any ideas I'd appreciate the help.

Note that the app not being shutdown properly causes a crash on lollipop (similar to how pre-lollipop is behaving now after this patch)

**Comment by est31** [comment]
[usenet has a thread on this](http://comments.gmane.org/gmane.comp.handhelds.android.ndk/11500). Everything else the internet spews out is similar. As it seems there is no good way to handle this.

**Comment by est31** [comment]
[this file](https://android.googlesource.com/platform/frameworks/base/+/master/services/core/java/com/android/server/am/ActivityStack.java) holds the source, as it seems the activity just has to be made "finishing".

**Comment by Zeno-** [comment]
@est31 Using exit() as suggested in the first link causes the problem...  Do you know how to properly implement what's in the second link?

**Comment by ShadowNinja** [comment]
I haven't looked closely at this, but the Minetest main thread seems to run in a separate native thread.  Maybe the Java thread isn't shutting down like it should?

File: src/porting_android.cpp [hunk 1]
```diff
@@ -88,6 +89,18 @@ android_app* app_global;
...
+
+void handleAndroidActivityEvents()
+{
+	int ident;
+	int events;
+	struct android_poll_source *source;
+
+	while ( (ident = ALooper_pollOnce(0, NULL, &events, (void**)&source)) >= 0)
+		if (source)
+			source->process(porting::app_global, source);
```
Comment (ShadowNinja) [inline src/porting_android.cpp]: At least the loop should have `{}`s.

**Comment by est31** [comment]
For reference: [some results from reading docs](http://irc.minetest.ru/minetest-dev/2015-05-06#i_4241652)

File: src/porting_android.cpp (hunk 1 above)
Comment (Zeno-) [inline src/porting_android.cpp]: Yep, you're probably right... I'll make an update when I read the docs @est31 has referenced

**Comment by Zeno-** [comment]
Thanks @est31... lots to read :)

**Comment by MoNTE48** [comment]
Note: On Lollipop 5.1 and 5.1.1 close minetest work good :)
Maybe, this android bug?

**Comment by est31** [comment]
The code in that file is still the same, also on 5.1. Are you sure that there is no error in the logcat?

**Comment by MoNTE48** [comment]
```
05-08 12:27:57.513  2928  2928 W ResourceType: No package identifier when getting value for resource number 0x00000000

05-08 12:27:57.514  2928  2928 W PackageManager: Failure retrieving resources for scd.lcex: Resource ID #0x0

05-08 12:27:57.524   606  3595 I ActivityManager: START u0 {act=com.android.systemui.recents.SHOW_RECENTS flg=0x10804000 cmp=com.android.systemui/.recents.RecentsActivity} from uid 10013 on display 0

05-08 12:27:57.649  2928  2951 W ResourceType: For resource 0x01030224, entry index(548) is beyond type entryCount(123)

05-08 12:27:57.651  2928  2951 W ResourceType: For resource 0x01030224, entry index(548) is beyond type entryCount(123)

05-08 12:27:57.655  3064  3084 W ResourceType: For resource 0x01030224, entry index(548) is beyond type entryCount(123)

05-08 12:27:57.656  3064  3084 W ResourceType: For resource 0x01030224, entry index(548) is beyond type entryCount(123)

05-08 12:27:57.858  2928  3635 W OpenGLRenderer: Incorrectly called buildLayer on View: TextView, destroying layer...

05-08 12:27:58.073  2928  2928 I Timeline: Timeline: Activity_idle id: android.os.BinderProxy@3e3e18a9 time:5507025

05-08 12:27:58.217   606   681 I Timeline: Timeline: Activity_windows_visible id: ActivityRecord{312d55f3 u0 com.android.systemui/.recents.RecentsActivity t123} time:5507169

05-08 12:27:58.578   606  3329 I ActivityManager: Config changes=480 {1.0 ?mcc?mnc ru_RU ?layoutDir sw600dp w960dp h527dp 320dpi lrg land finger -keyb/v/h -nav/h s.18 themeResource=themes:{default=overlay:com.pitchedapps.material.glass.free, iconPack:system, fontPkg:system, com.android.systemui=overlay:system, com.android.systemui.navbar=overlay:com.pitchedapps.material.glass.free}}

05-08 12:27:58.608   606  1276 I InputReader: Reconfiguring input devices.  changes=0x00000004

05-08 12:27:58.608   606  1276 I InputReader: Device reconfigured: id=6, name='elan-touchscreen', size 1200x1920, orientation 1, mode 1, display id 0

05-08 12:27:58.719 10652 10652 I Timeline: Timeline: Activity_idle id: android.os.BinderProxy@5f48093 time:5507672

05-08 12:27:58.734   606   681 I Timeline: Timeline: Activity_windows_visible id: ActivityRecord{37e1bfaa u0 net.minetest.minetest/.MtNativeActivity t136} time:5507686

05-08 12:27:58.952  2928  2928 W IInputConnectionWrapper: showStatusIcon on inactive InputConnection

05-08 12:27:58.987   606   681 I WindowManager: Screen frozen for +467ms due to Window{14e3692f u0 com.maxelus.shadowgalaxylivewallpaper.WallpaperService}

05-08 12:27:59.682  2928  2928 W ResourceType: No package identifier when getting value for resource number 0x00000000

05-08 12:27:59.682  2928  2928 W PackageManager: Failure retrieving resources for scd.lcex: Resource ID #0x0

05-08 12:28:00.217 10652 10677 I Minetest: Audio: Deinitializing...

05-08 12:28:00.237 10652 10677 I Minetest: Audio: Deinitialized.

05-08 12:28:00.238 10652 10677 I Minetest: GUIEngine: Deinitializing scripting

05-08 12:28:00.272 10652 10677 E Minetest: Shutting down minetest.

05-08 12:28:00.380   606  1305 I WindowState: WIN DEATH: Window{96b8767 u0 net.minetest.minetest/net.minetest.minetest.MtNativeActivity}

05-08 12:28:00.399   606  2880 I ActivityManager: Process net.minetest.minetest (pid 10652) has died

05-08 12:28:00.400   606  2880 W ActivityManager: Force removing ActivityRecord{37e1bfaa u0 net.minetest.minetest/.MtNativeActivity t136}: app died, no saved state

05-08 12:28:00.466   606  2880 I ActivityManager: Config changes=480 {1.0 ?mcc?mnc ru_RU ?layoutDir sw600dp w600dp h887dp 320dpi lrg port finger -keyb/v/h -nav/h s.19 themeResource=themes:{default=overlay:com.pitchedapps.material.glass.free, iconPack:system, fontPkg:system, com.android.systemui=overlay:system, com.android.systemui.navbar=overlay:com.pitchedapps.material.glass.free}}

05-08 12:28:00.466   606  1276 I InputReader: Reconfiguring input devices.  changes=0x00000004

05-08 12:28:00.466   606  1276 I InputReader: Device reconfigured: id=6, name='elan-touchscreen', size 1200x1920, orientation 0, mode 1, display id 0

05-08 12:28:01.070   606   681 I Timeline: Timeline: Activity_windows_visible id: ActivityRecord{22630163 u0 com.actionlauncher.playstore/com.actionlauncher.ActionLauncherActivity t110} time:5510022

05-08 12:28:01.076   606  3591 W InputMethodManagerService: Got RemoteException sending setActive(false) notification to pid 10652 uid 10094

05-08 12:28:01.238  6400  6400 I Timeline: Timeline: Activity_idle id: android.os.BinderProxy@26dac9b9 time:5510191

05-08 12:28:01.310   606   681 I WindowManager: Screen frozen for +902ms due to Window{14e3692f u0 com.maxelus.shadowgalaxylivewallpaper.WallpaperService}

05-08 12:28:01.621  6400  6470 W OpenGLRenderer: Incorrectly called buildLayer on View: ShortcutAndWidgetContainer, destroying layer...

05-08 12:28:01.621  6400  6470 W OpenGLRenderer: Incorrectly called buildLayer on View: ShortcutAndWidgetContainer, destroying layer...

05-08 12:28:02.995  6400  6400 I Timeline: Timeline: Activity_launch_request id:scd.lcex time:5511948

05-08 12:28:02.997   606  1305 I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 pkg=scd.lcex cmp=scd.lcex/.ActivityMain bnds=[200,74][400,377]} from uid 10090 on display 0

05-08 12:28:03.000   606  1305 W ResourceType: For resource 0x01030224, entry index(548) is beyond type entryCount(123)

05-08 12:28:03.001   606  1305 W ResourceType: For resource 0x01030224, entry index(548) is beyond type entryCount(123)

05-08 12:28:03.168   606   681 W ResourceType: For resource 0x01030224, entry index(548) is beyond type entryCount(123)

05-08 12:28:03.169   606   681 W ResourceType: For resource 0x01030224, entry index(548) is beyond type entryCount(123)

05-08 12:28:03.198  2928  3635 W OpenGLRenderer: Incorrectly called buildLayer on View: TextView, destroying layer...

05-08 12:28:03.262 11206 11206 I Timeline: Timeline: Activity_idle id: android.os.BinderProxy@335d2aec time:5512215

05-08 12:28:03.619   606   681 I Timeline: Timeline: Activity_windows_visible id: ActivityRecord{2771449d u0 scd.lcex/.ActivityMain t140} time:5512572

```
On reopen game no problem, all work.
CyanogenMod 12.1 (Android 5.1.1)

**Comment by est31** [comment]
So you do experience the issue Zeno tries to fix here. The restart problem is of his current fix.

**Comment by MoNTE48** [comment]
Yes, on 5.0 I see this problem. On 5.1 all good

**Comment by est31** [comment]
Interesting, behaviour changed. We'll still have to support older android versions.

**Comment by Zeno-** [comment]
This doesn't fix the problem. I will leave the PR open, though, remove the unnecessary stuff and when finished rename it to just cleanup shutdown sequence

**Comment by Zeno-** [comment]
Will clean this up and make a new PR

**Pull request closed without merging**